### PR TITLE
fix: re-enable filter related rpc

### PR DESCRIPTION
**Pull request by RetricSu**

### CODE REVIEW

File: packages/api-server/src/db/index.ts [hunk 1]
```diff
@@ -170,6 +183,124 @@ export class Query {
...
     return [formatTransaction(tx), logs.map((log) => formatLog(log))];
   }
+
+  private async queryLogsByBlockHash(
+    blockHash: HexString,
+    address?: HexString,
+    lastPollId?: string
+  ): Promise<Log[]> {
+    const queryAddress = address ? { address } : {};
+    const queryLastPollId = lastPollId ? lastPollId : -1;
```
Comment (classicalliu) [inline packages/api-server/src/db/index.ts:193]: `lastPollId` is string or number ?

File: packages/api-server/src/db/index.ts [hunk 2]
```diff
@@ -170,6 +183,124 @@ export class Query {
...
+    let logs = await this.knex
+      .select()
+      .table("logs")
```
Comment (classicalliu) [inline packages/api-server/src/db/index.ts:196]:
```suggestion
    let logs = await this.knex<Log>("logs")
```

File: packages/api-server/src/db/index.ts [hunk 3]
```diff
@@ -44,6 +46,17 @@ export class Query {
     return formatBlock(block);
   }
 
+  async getBlocksAfterBlockNumber(
+    number: BigInt,
+    _order?: "desc" | "asc"
+  ): Promise<Block[]> {
+    const order = _order || "desc";
```
Comment (classicalliu) [inline packages/api-server/src/db/index.ts:53]:
```suggestion
    order?: "desc" | "asc" = "desc"
  ): Promise<Block[]> {
```

File: packages/api-server/src/db/index.ts [hunk 4]
```diff
@@ -170,6 +183,124 @@ export class Query {
...
+    fromBlock: HexNumber,
+    toBlock: HexNumber,
+    address?: HexString,
+    lastPollId?: string
+  ): Promise<Log[]> {
+    const queryAddress = address ? { address } : {};
+    const queryLastPollId = lastPollId ? lastPollId : -1;
+    let logs = await this.knex
+      .select()
+      .table("logs")
```
Comment (classicalliu) [inline packages/api-server/src/db/index.ts:215]:
```suggestion
    let logs = await this.knex<Log>("logs")
```

File: packages/api-server/src/db/index.ts [hunk 5]
```diff
@@ -170,6 +183,124 @@ export class Query {
...
+    lastPollId: number,
+    option: LogQueryOption,
+    fromBlock: BigInt,
+    toBlock: BigInt
+  ): Promise<Log[]>;
+
+  async getLogsAfterLastPoll(
+    lastPollId: number,
+    option: LogQueryOption,
+    blockHashOrFromBlock: HexString | BigInt,
```
Comment (classicalliu) [inline packages/api-server/src/db/index.ts:277]:
```suggestion
    blockHashOrFromBlock: HexString | bigint,
```

File: packages/api-server/src/db/index.ts [hunk 6]
```diff
@@ -170,6 +183,124 @@ export class Query {
...
+    toBlock?: BigInt
```
Comment (classicalliu) [inline packages/api-server/src/db/index.ts:278]:
```suggestion
    toBlock?: bigint
```

File: packages/api-server/src/db/index.ts [hunk 7]
```diff
@@ -44,6 +46,17 @@ export class Query {
     return formatBlock(block);
   }
 
+  async getBlocksAfterBlockNumber(
+    number: BigInt,
```
Comment (classicalliu) [inline packages/api-server/src/db/index.ts:50]:
```suggestion
    number: bigint,
```

File: packages/api-server/src/db/index.ts [hunk 8]
```diff
@@ -170,6 +183,124 @@ export class Query {
...
+      .where("id", ">", queryLastPollId)
+      .orderBy("id", "desc");
+    logs = logs.map((log) => formatLog(log));
+    return logs;
+  }
+
+  getLogs(
+    option: LogQueryOption,
+    fromBlock: BigInt,
+    toBlock: BigInt
```
Comment (classicalliu) [inline packages/api-server/src/db/index.ts:228]:
```suggestion
    fromBlock: bigint,
    toBlock: bigint
```

File: packages/api-server/src/db/index.ts [hunk 9]
```diff
@@ -216,3 +347,98 @@ function toBigIntOpt(num: bigint | HexNumber | undefined): bigint | undefined {
...
+    if (!match) {
+      continue;
+    }
+    result.push(log);
+  }
+  return result;
+}
+
+// test
+export async function test_topic_match() {
```
Comment (classicalliu) [inline packages/api-server/src/db/index.ts:414]:
```suggestion
export async function testTopicMatch() {
```

File: packages/api-server/src/methods/modules/eth.ts [hunk 10]
```diff
@@ -618,45 +624,171 @@ export class Eth {
...
-  async getFilterLogs(args: [string]): Promise<void> {
-    throw new MethodNotSupportError("eth_getFilterLogs is not supported!");
+  async getFilterLogs(args: [string]): Promise<Array<any>> {
+    const filter_id = parseInt(args[0], 16);
+    const filter = this.filterManager.get(filter_id);
+
+    if (!filter)
+      throw new Web3Error(
+        `invalid filter id ${filter_id}. the filter might be removed or outdated.`
+      );
```
Comment (classicalliu) [inline packages/api-server/src/methods/modules/eth.ts:660]:
```suggestion
    if (!filter) {
      throw new Web3Error(
        `invalid filter id ${filter_id}. the filter might be removed or outdated.`
      );
    }
```

File: packages/api-server/src/methods/modules/eth.ts [hunk 11]
```diff
@@ -618,45 +624,171 @@ export class Eth {
...
-  async getFilterChanges(args: [string]): Promise<void> {
-    throw new MethodNotSupportError("eth_getFilterChanges is not supported!");
+  async getFilterChanges(args: [string]): Promise<string[] | EthLog[]> {
+    const filter_id = parseInt(args[0], 16);
+    const filter = this.filterManager.get(filter_id);
+
+    if (!filter)
+      throw new Web3Error(
+        `invalid filter id ${filter_id}. the filter might be removed or outdated.`
+      );
```
Comment (classicalliu) [inline packages/api-server/src/methods/modules/eth.ts:688]:
```suggestion
    if (!filter) {
      throw new Web3Error(
        `invalid filter id ${filter_id}. the filter might be removed or outdated.`
      );
    }
```

File: packages/api-server/src/methods/modules/eth.ts [hunk 12]
```diff
@@ -618,45 +624,171 @@ export class Eth {
...
+      // remember to update the last poll cache
+      // logsData[0] is now the highest log id(meaning it is the newest cache log id)
+      this.filterManager.updateLastPoll(filter_id, Number(logs[0].id));
+      return logs.map((log) => toApiLog(log));
+    }
+
+    const fromBlockNumber: U64 = await this.blockParameterToBlockNumber(
+      filter.fromBlock || "latest"
+    );
+    const toBlockNumber: U64 | undefined =
```
Comment (classicalliu) [inline packages/api-server/src/methods/modules/eth.ts:741]:
```suggestion
    const toBlockNumber: U64 =
```

File: packages/api-server/src/methods/modules/eth.ts [hunk 13]
```diff
@@ -618,45 +624,171 @@ export class Eth {
...
+    // if blockHash exits, fromBlock and toBlock is not allowed.
+    if (blockHash) {
+      const logs = await this.query.getLogs(queryOption, blockHash);
+      return logs.map((log) => toApiLog(log));
+    }
+
+    const fromBlockNumber: U64 = await this.blockParameterToBlockNumber(
+      filter.fromBlock || "latest"
+    );
+    const toBlockNumber: U64 | undefined =
```
Comment (classicalliu) [inline packages/api-server/src/methods/modules/eth.ts:778]:
```suggestion
    const toBlockNumber: U64 =
```

File: packages/api-server/src/db/index.ts (hunk 1 above)
Comment (RetricSu) [reply packages/api-server/src/db/index.ts]: should be number. fixed.

File: packages/api-server/src/db/index.ts [hunk 14]
```diff
@@ -170,6 +182,120 @@ export class Query {
...
+      .orderBy("id", "desc");
+    logs = logs.map((log) => formatLog(log));
+    return logs;
+  }
+
+  private async queryLogsByBlockRange(
+    fromBlock: HexNumber,
+    toBlock: HexNumber,
+    address?: HexString,
+    lastPollId?: string
```
Comment (classicalliu) [inline packages/api-server/src/db/index.ts:206]:
```suggestion
    lastPollId?: number
```

**Review by classicalliu — APPROVED** [review]

**Pull request merged**